File: src/org/lineageos/xiaomi_bluetooth/earbuds/Earbud.java
package org.lineageos.xiaomi_bluetooth.earbuds;

import androidx.annotation.NonNull;


public class Earbud {

    private static final int EARBUDS_CHARGING_BIT_MASK = 0x80;
    private static final int EARBUDS_BATTERY_LEVEL_MASK = 0x7F;

    public final boolean charging;
    public final int battery;

    public Earbud(byte raw) {
        this.charging = (raw & EARBUDS_CHARGING_BIT_MASK) != 0;
        this.battery = raw & EARBUDS_BATTERY_LEVEL_MASK;
    }

    public boolean isValid() {
        return this.battery <= 100;
    }

    @NonNull
    @Override
    public String toString() {
        return "Earbud{" +
                "charging=" + charging +
                ", battery=" + battery +
                '}';
    }
}

File: src/org/lineageos/xiaomi_tws/earbuds/Earbud.kt
package org.lineageos.xiaomi_tws.earbuds

import com.android.settingslib.bluetooth.BluetoothUtils.META_INT_ERROR

data class Earbud(val raw: Byte) {

    val charging: Boolean
    val battery: Int
    val valid: Boolean

    init {
        val battery = raw.toInt() and EARBUDS_BATTERY_LEVEL_MASK
        val charging = (raw.toInt() and EARBUDS_CHARGING_BIT_MASK) != 0
        val valid = battery <= 100

        this.battery = if (valid) battery else META_INT_ERROR
        this.charging = charging and valid
        this.valid = valid
    }

    override fun toString() = "Earbud{charging=$charging, battery=$battery, valid=$valid}";

    companion object {
        private const val EARBUDS_CHARGING_BIT_MASK = 0x80
        private const val EARBUDS_BATTERY_LEVEL_MASK = 0x7F
    }
}
